bolt-boost/src/constraints.rs

```diff
@@ -0,0 +1,87 @@
+use std::{
+    collections::HashMap,
+    sync::{Arc, RwLock},
+};
+use tracing::error;
+
+use super::types::{ConstraintsMessage, ConstraintsWithProofData};
+
+/// A concurrent cache of constraints.
+#[derive(Clone, Default, Debug)]
+pub struct ConstraintsCache {
+    cache: Arc<RwLock<HashMap<u64, Vec<ConstraintsWithProofData>>>>,
+}
+
+impl ConstraintsCache {
+    pub fn new() -> Self {
+        Self {
+            cache: Default::default(),
+        }
+    }
+
+    /// Checks if the constraints for the given slot conflict with the existing constraints.
+    /// Will check for:
+    /// - Multiple ToB constraints per slot
+    /// - Duplicates of the same transaction per slot
+    pub fn conflicts_with(&self, slot: &u64, constraints: &ConstraintsMessage) -> bool {
+        if let Some(saved_constraints) = self.cache.read().unwrap().get(slot) {
+            for saved_constraint in saved_constraints {
+                // Only 1 ToB constraint per slot
+                if constraints.top && saved_constraint.message.top {
+                    return true;
+                }
+
+                // Check if the transactions are the same
+                for tx in &constraints.transactions {
+                    if saved_constraint
+                        .message
+                        .transactions
+                        .iter()
+                        .any(|existing| tx == existing)
+                    {
+                        return true;
+                    }
+                }
+            }
+        }
+
+        false
+    }
+
+    /// Inserts the constraints for the given slot. Also decodes the raw transactions to save their
+    /// transaction hashes and hash tree roots for later use. Will first check for conflicts, and return
+    /// false if there are any.
+    ///
+    /// TODO: return Result instead of bool
```

Comment:
TODO left here: do you want to address this in a future PR?

Reply:
Thanks for the reminder, done